[PowerPC][X86] Make cpu id builtins target independent and lower for PPC

nemanjai opened the pull request:
Make __builtin_cpu_{init|supports|is} target independent and provide an opt-in query for targets that want to support it. Each target is still responsible for their specific lowering/code-gen. Also provide code-gen for PowerPC.

I originally proposed this in https://reviews.llvm.org/D152914 and this addresses the comments I received there.

Commit:
[PowerPC][X86] Make cpu id builtins target independent and lower for PPC

Make __builtin_cpu_{init|supports|is} target independent and provide
an opt-in query for targets that want to support it. Each target is
still responsible for their specific lowering/code-gen.
Also provide code-gen for PowerPC.

Commit:
Run clang-format which I appear to have forgotten to run

ilinpv commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2110,6 +2110,66 @@ static bool checkFPMathBuiltinElementType(Sema &S, SourceLocation Loc,
+  if (TI.supportsCpuSupports())
+    TheTI = &TI;
+  else if (AuxTI && AuxTI->supportsCpuSupports())
+    TheTI = AuxTI;
```
If these builtins are needed for auxiliary target ( are they ? ) it would good to have tests for that.

ilinpv reviewed (commented):
Except for the AuxTarget question, target independent part looks good to me. Thank you for the patch.

kamaub commented on clang/test/CodeGen/builtin-cpu-supports.c:
```diff
@@ -1,11 +1,16 @@
+// RUN: %clang_cc1 -triple ppc64le-linux-gnu -emit-llvm < %s | FileCheck %s \
+// RUN:   --check-prefix=CHECK-PPC
+
+#ifndef __PPC__
```
why not use `// REQUIRES: x86-registered-target`

amy-kwan replied:
It might be better to have the PPC checks in a separate file from the x86 RUN lines so we don't need to do the `ifndef`/`else`.

kamaub commented on llvm/include/llvm/IR/IntrinsicsPowerPC.td:
```diff
@@ -210,6 +210,12 @@ let TargetPrefix = "ppc" in {  // All intrinsics start with "llvm.ppc.".
             [llvm_float_ty],
             [llvm_float_ty, llvm_float_ty, llvm_float_ty, llvm_vararg_ty],
             [IntrNoMem]>;
+  // Load of a value provided by the system library at a fixed address. Used for
+  // accessing things like HWCAP word provided by GLIBC.
```
```suggestion
// This intrinsic is provided to allow back ends to emit load
// instructions that load a value from a fixed address. The
// parameter to the intrinsic is not an address, but an
// immediate index into an enumeration that contains the
// union of all such values available on all back ends.
// An example is the HWCAP/HWCAP2/CPUID words
// provided by GLIBC on PowerPC to allow fast access
// to commonly used parts of AUXV. These are provided
// at a fixed offset into the TCB (accessible through the
// thread pointer).
```
Missing expanded comment from previous round of reviews in https://reviews.llvm.org/D152914

kamaub commented on llvm/include/llvm/TargetParser/PPCTargetParser.def:
```diff
@@ -0,0 +1,80 @@
+#ifndef PPC_FEATURE
```
Missing file description similar to the `head` of `clang/include/clang/Basic/Builtins.def`

diggerlin replied:
we will support these feature in AIX OS soon.
can we change `PPC_FEATURE` to `PPC_LINUX_FEATURE`

and `PPC_CPU` to `PPC_LINUX_CPU` 

and when implement the __builtin_cpu_is and __builtin_cpu_supports in AIX OD

we can `PPC_AIX_FEATURE` and `PPC_AIX_CPU`

amy-kwan commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2110,6 +2110,66 @@ static bool checkFPMathBuiltinElementType(Sema &S, SourceLocation Loc,
+           << SourceRange(TheCall->getBeginLoc(), TheCall->getEndLoc());
+
+  // Check if the argument is a string literal.
+  if (!isa<StringLiteral>(Arg->IgnoreParenImpCasts()))
```
It seems like we can pull out some of the calls are are repeated, such as `Arg->IgnoreParenImpCasts()`.

amy-kwan commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2110,6 +2110,66 @@ static bool checkFPMathBuiltinElementType(Sema &S, SourceLocation Loc,
+           << SourceRange(TheCall->getBeginLoc(), TheCall->getEndLoc());
+
+  // Check if the argument is a string literal.
+  if (!isa<StringLiteral>(Arg->IgnoreParenImpCasts()))
```
Likewise here, perhaps we can pull out some of the calls that we call more than once.

amy-kwan reviewed (commented):
Initial group code review comments.

efriedma-quic commented on clang/lib/Basic/Targets/PPC.h:
```diff
@@ -359,6 +359,13 @@ class LLVM_LIBRARY_VISIBILITY PPCTargetInfo : public TargetInfo {
     return RegName.equals("r1") || RegName.equals("x1");
   }
+
+  // We support __builtin_cpu_supports/__builtin_cpu_is on targets that
```
Musl targets are supposed to use a "musl" triple.  We already have Triple::isMusl(). If it's somehow possible for both isOSGlibc() and isMusl() to be true, I'd consider that a bug in isOSGlibc().

MaskRay replied:
Linux musl also supports `AT_HWCAP` `AT_HWCAP2`.

nemanjai replied:
We can certainly expand this in a follow-up patch, but since I don't really have a way to do any functional testing on a musl machine, I'll leave it out of this patch.

amy-kwan reviewed (commented):
Additional group code review comments.

diggerlin commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2110,6 +2110,66 @@ static bool checkFPMathBuiltinElementType(Sema &S, SourceLocation Loc,
+/// SemaBuiltinCpuSupports - Handle __builtin_cpu_supports(char *).
+/// This checks that the target supports __builtin_cpu_supports and
+/// that the string argument is constant and valid.
+static bool SemaBuiltinCpuSupports(Sema &S, const TargetInfo &TI,
```
most of the source code of function `SemaBuiltinCpuSupports` and  `SemaBuiltinCpuIs` are same , we can add helper function to reduce the code.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCTargetMachine.h:
```diff
@@ -32,6 +32,7 @@ class PPCTargetMachine final : public LLVMTargetMachine {
   std::unique_ptr<TargetLoweringObjectFile> TLOF;
   PPCABI TargetABI;
   Endian Endianness = Endian::NOT_DETECTED;
+  mutable bool HasGlibcHWCAPAccess = false;
```
Question: Why does this need to be mutable?

amy-kwan commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -1830,6 +1830,10 @@ void PPCLinuxAsmPrinter::emitEndOfAsmFile(Module &M) {
   PPCTargetStreamer *TS =
       static_cast<PPCTargetStreamer *>(OutStreamer->getTargetStreamer());
 
+  if (static_cast<const PPCTargetMachine &>(TM).hasGlibcHWCAPAccess())
```
Can we add more documentation regarding what we're doing here?

amy-kwan reviewed (commented):
Additional group review comments.

amy-kwan commented on clang/test/CodeGen/builtin-cpu-supports.c:
```diff
@@ -15,38 +20,57 @@ int main(void) {
+// CHECK-X86-NEXT:    and i32 [[LOAD]], 4
 int v4() { return __builtin_cpu_supports("x86-64-v4"); }
+#else
+int test(int a) {
```
Minor nit: Can we add a `CHECK-LABEL` as well? So that we know that we're actually inside `test`.

amy-kwan reviewed (commented):
Additional group review comments.

Commit:
Address comments

- Use script for generating test checks
- Clean up code and rename macros for clarity
- Remove magic number offsets
- Get rid of auxiliary targets in Sema checking
- Clarify some comments and add missing ones
- Add check for undefined symbol added for safety

nemanjai commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2110,6 +2110,66 @@ static bool checkFPMathBuiltinElementType(Sema &S, SourceLocation Loc,
+/// This checks that the target supports __builtin_cpu_supports and
+/// that the string argument is constant and valid.
+static bool SemaBuiltinCpuSupports(Sema &S, const TargetInfo &TI,
+                                   const TargetInfo *AuxTI, CallExpr *TheCall) {
```
Turns out this is actually needed. There is a test that builds with `-triple aarch64 -aux-triple x86_64`. I will have to put the AUX target checks back.

Commit:
Add the Auxiliary target checks back in for Sema.

diggerlin commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -16086,6 +16086,41 @@ Value *CodeGenFunction::EmitPPCBuiltinExpr(unsigned BuiltinID,
+#define PPC_LNX_FEATURE(Name, Description, EnumName, Bitmask, FA_WORD)         \
+  .Case(Name, {FA_WORD, Bitmask})
+#include "llvm/TargetParser/PPCTargetParser.def"
+            .Default({0, 0});
```
we do not want a default here,  without a default it will hit an assert here if fell off the end of the string-switch.
in the class of StringSwitch , there is function as 

```
  [[nodiscard]] operator R() {
    assert(Result && "Fell off the end of a string-switch");
    return std::move(*Result);
  }
```

nemanjai replied:
I think this is a good point. If we somehow have a string argument that would produce the `default` case and it has made it past Sema checking, it would be good to crash here rather than produce an invalid call to the intrinsic (or an invalid mask).
Of course, that assert is not all that friendly and it might be better to assert within this function.

What do you think about keeping the `default` and adding an assert below such as:
```
assert(BitMask && "Invalid target feature string. Missed by SemaChecking?");
```

amy-kwan commented on llvm/include/llvm/IR/IntrinsicsPowerPC.td:
```diff
@@ -210,6 +210,15 @@ let TargetPrefix = "ppc" in {  // All intrinsics start with "llvm.ppc.".
             [llvm_float_ty, llvm_float_ty, llvm_float_ty, llvm_vararg_ty],
             [IntrNoMem]>;
+  // Load of a value provided by the system library at a fixed address. Used for
+  // accessing things like HWCAP word provided by GLIBC. The immediate argument
```
Minor nits:
- Should it be `the HWCAP word`?
- We use GLIBC on this line but Glibc on line 217. I think it would be better if we were consistent and use one way of addressing GLIBC.

nemanjai replied:
I'm fine with both of these suggestions.

amy-kwan reviewed (commented):
I think I primarily have minor nit comments. Thank you for the update, Nemanja.

amy-kwan reviewed (approved):
I meant to approve this before but forgot. I think LGTM once the comments are addressed.

nemanjai committed:
Address last round of reviews and rebase

The way we define target-independent builtins has changed which
required a rebase. Also cleaned up some comments and improved
asserts as suggested in the review.

Aokamal commented:
Hi, 

Sorry I kind of don't know much about the change. We have the following code snippet
```
#if ABSL_HAVE_BUILTIN(__builtin_cpu_supports)
  if (__builtin_cpu_supports("avx2")) {
    ...
  }
#endif
```

when run with arm cpu, it fails with `error: builtin is not supported on this target` on the second if. Does it mean  ABSL_HAVE_BUILTIN is not accurate here ?

lei137 commented:
Hi Aokamal,

This maco was not taken into account in the implementation. Can you please open a bug and assign the issue to nemanjai?

Thank-you for reporting this.

Regards,
Lei

From: Aokamal ***@***.***>
Date: Tuesday, January 30, 2024 at 10:30 AM
To: llvm/llvm-project ***@***.***>
Cc: Lei Huang ***@***.***>, State change ***@***.***>
Subject: [EXTERNAL] Re: [llvm/llvm-project] [PowerPC][X86] Make cpu id builtins target independent and lower for PPC (PR #68919)
Hi, Sorry I kind of don't know much about the change. We have the following code snippet #if ABSL_HAVE_BUILTIN(__builtin_cpu_supports) if (__builtin_cpu_supports("avx2")) { .. . } #endif when run with arm cpu, it fails with error: builtin is
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2U-qSTa06VIRMu4YcIkOZYJv87-YdAlWKJZhcdfdL-hOL9LK6PP42RfIRBxJiqbiimkckMLPKtrGCex3SbDVRbwy2koA67SnPo8$>
Report Suspicious <https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2U-qSTa06VIRMu4YcIkOZYJv87-YdAlWKJZhcdfdL-hOL9LK6PP42RfIRBxJiqbiimkckMLPKtrGCex3SbDVRbwy2koA67SnPo8$>


ZjQcmQRYFpfptBannerEnd

Hi,

Sorry I kind of don't know much about the change. We have the following code snippet

#if ABSL_HAVE_BUILTIN(__builtin_cpu_supports)

  if (__builtin_cpu_supports("avx2")) {

    ...

  }

#endif

when run with arm cpu, it fails with error: builtin is not supported on this target on the second if. Does it mean ABSL_HAVE_BUILTIN is not accurate here ?

—
Reply to this email directly, view it on GitHub<https://github.com/llvm/llvm-project/pull/68919#issuecomment-1917219523>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKTLOI65B4HXCU6BOZW3G73YREGYJAVCNFSM6AAAAAA56ALXCOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJXGIYTSNJSGM>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

nemanjai commented:
Sorry, I'll put up a patch for review shortly to fix this problem.
